### ISSUE TITLE
 uninitialized constant Mysql2::Client::REMEMBER_OPTIONS error with newest ruby and mysql2 versions on Windows7 (MariaDB)

### ISSUE THREAD

**Issue by LudwigKrause** [brianmario/mysql2#954]
Hi,

I get the following error when trying to run this simple ruby-file:
```
require 'mysql2'
client = Mysql2::Client.new(:host => "myhost", :username=> "root")
```
Error:

```
Traceback (most recent call last):
        3: from D:/RubyMySql/sqltest.rb:2:in `<main>'
        2: from D:/RubyMySql/sqltest.rb:2:in `new'
        1: from C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/mysql2-0.5.0-x64-mingw32/lib/mysql2/client.rb:25:in `initialize'
C:/Ruby25-x64/lib/ruby/gems/2.5.0/gems/mysql2-0.5.0-x64-mingw32/lib/mysql2/client.rb:14:in `default_query_options': uninitialized constant Mysql2::Client::REM
EMBER_OPTIONS (NameError)
```

My Setup is: 

- mysql2 0.5.0
- gem 2.7.3
- ruby 2.5
- Windows 7 
- MariaDB Connector c 3.0.3 (also tried full MariaDB Installation version 10.3.5, same error)

I installed mysql2 from the cmd window as follows:

> gem install mysql2 -- --with-mysql-dir="C:/Program Files/MariaDB/MariaDB Connector C 64-bit"

Did I do something wrong or is there an issue with this setup?

**Comment by sodabrew** [comment]
The header value is defined by the MariaDB Connector/C 3.0 headers: https://github.com/MariaDB/mariadb-connector-c/blob/3.0/include/mariadb_com.h#L164

Since the package compiles at all, and your command line looks right, I am not sure why the code missed out of this definition.

**Comment by aldent95** [comment]
Getting the same issue.
Windows 10
Rails: 5.0.7
Ruby: 2.4.2p198
Mysql2: 0.5.0
Normal Mysql2 installation

Full stacktrace:

```
NameError (uninitialized constant Mysql2::Client::REMEMBER_OPTIONS):
  
mysql2-0.5.0-x64 (mingw32) lib/mysql2/client.rb:14:in `default_query_options'
mysql2-0.5.0-x64 (mingw32) lib/mysql2/client.rb:25:in `initialize'
activerecord (5.0.7) lib/active_record/connection_adapters/mysql2_adapter.rb:25:in `new'
activerecord (5.0.7) lib/active_record/connection_adapters/mysql2_adapter.rb:25:in `mysql2_connection'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:729:in `new_connection'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:773:in `checkout_new_connection'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:752:in `try_to_checkout_new_connection'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:713:in `acquire_connection'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:490:in `checkout'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:364:in `connection'
activerecord (5.0.7) lib/active_record/connection_adapters/abstract/connection_pool.rb:883:in `retrieve_connection'
activerecord (5.0.7) lib/active_record/connection_handling.rb:128:in `retrieve_connection'
activerecord (5.0.7) lib/active_record/connection_handling.rb:91:in `connection'
activerecord (5.0.7) lib/active_record/migration.rb:559:in `connection'
activerecord (5.0.7) lib/active_record/migration.rb:546:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/callbacks.rb:38:in `block in call'
activesupport (5.0.7) lib/active_support/callbacks.rb:97:in `__run_callbacks__'
activesupport (5.0.7) lib/active_support/callbacks.rb:750:in `_run_call_callbacks'
activesupport (5.0.7) lib/active_support/callbacks.rb:90:in `run_callbacks'
actionpack (5.0.7) lib/action_dispatch/middleware/callbacks.rb:36:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/remote_ip.rb:79:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/debug_exceptions.rb:49:in `call'
web-console (2.3.0) lib/web_console/middleware.rb:28:in `block in call'
web-console (2.3.0) lib/web_console/middleware.rb:18:in `catch'
web-console (2.3.0) lib/web_console/middleware.rb:18:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.0.7) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.0.7) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.0.7) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.0.7) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.0.7) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.0.7) lib/rails/rack/logger.rb:24:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.4) lib/rack/method_override.rb:22:in `call'
rack (2.0.4) lib/rack/runtime.rb:22:in `call'
activesupport (5.0.7) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.7) lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.4) lib/rack/sendfile.rb:111:in `call'
railties (5.0.7) lib/rails/engine.rb:522:in `call'
puma (3.11.3) lib/puma/configuration.rb:225:in `call'
puma (3.11.3) lib/puma/server.rb:624:in `handle_request'
puma (3.11.3) lib/puma/server.rb:438:in `process_client'
puma (3.11.3) lib/puma/server.rb:302:in `block in run'
puma (3.11.3) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/_source.html.erb
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/_source.html.erb (1.0ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.0ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.0ms)
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/actionpack-5.0.7/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout (697.9ms)
DEPRECATION WARNING: Accessing mime types via constants is deprecated. Please change `Mime::HTML` to `Mime[:html]`. (called from acceptable_content_type? at C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/middleware.rb:58)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/index.html.erb
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/_markup.html.erb (0.0ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/console.js.erb within layouts/javascript
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/_inner_console_markup.html.erb within layouts/inlined_string
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/_inner_console_markup.html.erb within layouts/inlined_string (0.5ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/_prompt_box_markup.html.erb within layouts/inlined_string
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/_prompt_box_markup.html.erb within layouts/inlined_string (0.5ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/style.css.erb within layouts/inlined_string
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/style.css.erb within layouts/inlined_string (0.0ms)
DEPRECATION WARNING: Accessing mime types via constants is deprecated. Please change `Mime::WEB_CONSOLE_V2` to `Mime[:web_console_v2]`. (called from _____uby___x___lib_ruby_gems_______gems_web_console_______lib_web_console_templates_console_js_erb___1970265061_105767840 at C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/console.js.erb:491)
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/console.js.erb within layouts/javascript (633.1ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/main.js.erb within layouts/javascript
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/main.js.erb within layouts/javascript (0.5ms)
  Rendering C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/error_page.js.erb within layouts/javascript
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/error_page.js.erb within layouts/javascript (0.5ms)
  Rendered C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/web-console-2.3.0/lib/web_console/templates/index.html.erb (1353.3ms)
```

**Comment by reklawevarg** [comment]
Same issue, Windows 10, Ruby  2.3.3p222 (2016-11-21 revision 56859) [x64-mingw32]

Reverting to mysql2 (0.4.9 x64-mingw32) resolves the issue.

**Comment by maiken2051** [comment]
For posterity, I got mysql2 0.4.10 working with Ruby 2.4.4 by downloading the MySQL connector from:

https://dev.mysql.com/downloads/connector/c/

I expanded the Zip file into my Ruby directory, then I built the gem:

`gem install mysql2 --version 0.4.10 --platform=ruby -- '--with-mysql-lib="C:\Ruby24\mysql-connector-c-6.1.11-winx64\lib" --with-mysql-include="C:\Ruby24\mysql-connector-c-6.1.11-winx64\include"'`

Tip of the hat to [#861](https://github.com/brianmario/mysql2/issues/861) which helped me figure out how to build the gem.

**Comment by sodabrew** [comment]
Please try mysql2 0.5.1 binaries now.

**Comment by LudwigKrause** [comment]
Yes. It works now. Thank your very much @sodabrew for your quick fix. I'm closing the issue now

**Comment by BMahabir** [comment]
@sodabrew @LudwigKrause wait, are you guys saying I can now use mysql 0.5.x with ruby 2.4.x in Windows without issue anymore?  This is huge if so.

**Comment by sodabrew** [comment]
Yes, the binary builds of mysql2 gem 0.5.x include Ruby 2.0 - 2.5 inclusive.

**Comment by BMahabir** [comment]
OK so I tried it, and I got the same error that I always had before : 

C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/activerecord-4.2.10/lib/active_record/connection_adapters/connection_specification.rb:177:in `rescue in spec': Specified 'mysql2' for database adapter, but the gem is not loaded. Add `gem 'mysql2'` to your Gemfile (and ensure its version is at the minimum required by ActiveRecord). (Gem::LoadError)

Am I missing something here?
